File: src/metamodel_builder.py
"""
Meta-Model Builder - Heuristic-based world generation
Creates meta-models and conceptual worlds from extracted concepts
Uses graph algorithms and procedural generation without LLMs
"""

import json
import math
from typing import List, Dict, Tuple, Any
from collections import defaultdict
import numpy as np


class MetaModelBuilder:
    """
    Builds meta-models and conceptual worlds using heuristic algorithms.
    Implements multidimensional space generation and concept organization.
    """

    def __init__(self, dimensions: int = 3):
        """
        Initialize the meta-model builder.

        Args:
            dimensions: Number of dimensions for the conceptual space
        """
        self.dimensions = dimensions
        self.concept_positions = {}
        self.layers = []

    def create_singularity(self, task: str, concept_graph: Dict) -> Dict:
        """
        Create a Singularity - the core semantic node.

        Args:
            task: The task or idea description
            concept_graph: Graph of extracted concepts

        Returns:
            Singularity data structure
        """
        # Calculate layers based on concept complexity
        layers_count = self._calculate_layers(concept_graph)

        # Generate visual representation
        visual_rep = self._generate_visual_representation(concept_graph)

        singularity = {
            'task': task,
            'task_vector': self._text_to_vector(task),
            'layers_count': layers_count,
            'visual_representation': visual_rep,
            'content': json.dumps(concept_graph),
            'version': 1
        }

        return singularity

    def create_points(self, concept_graph: Dict, singularity_id: int = None) -> List[Dict]:
        """
        Create Points from concept graph nodes.

        Args:
            concept_graph: Graph of extracted concepts
            singularity_id: ID of parent singularity

        Returns:
            List of Point data structures
        """
        points = []

        for node in concept_graph.get('nodes', []):
            # Generate position vector in multidimensional space
            position_vector = self._generate_position_vector(node)

            # Determine point type based on node characteristics
            point_type = self._determine_point_type(node)

            # Calculate layer based on node weight and connections
            layer = self._calculate_point_layer(node, concept_graph)

            point = {
                'task': node.get('id', ''),
                'task_vector': self._text_to_vector(node.get('id', '')),
                'position_vector': position_vector,
                'layer': layer,
                'space_dimension': self.dimensions,
                'name': node.get('id', ''),
                'content': json.dumps(node),
                'weight': node.get('weight', 1.0),
                'type': point_type,
                'is_super_cluster': node.get('weight', 0) > 0.8,
                'visual_representation': self._generate_point_visualization(node, position_vector),
                'singularity_id': singularity_id
            }

            points.append(point)
            self.concept_positions[node['id']] = position_vector

        return points

    def create_segments(self, concept_graph: Dict, points: List[Dict]) -> List[Dict]:
        """
        Create Segments (connections between Points).

        Args:
            concept_graph: Graph of extracted concepts
            points: List of created points

        Returns:
            List of Segment data structures
        """
        segments = []
        point_id_map = {p['name']: i for i, p in enumerate(points)}

        for edge in concept_graph.get('edges', []):
            from_node = edge.get('from')
            to_node = edge.get('to')

            if from_node in point_id_map and to_node in point_id_map:
                segment = {
                    'type_name': edge.get('type', 'connection'),
                    'type': self._encode_segment_type(edge.get('type', 'connection')),
                    'weight': edge.get('weight', 1.0),
                    'from_point_idx': point_id_map[from_node],
                    'to_point_idx': point_id_map[to_node]
                }
                segments.append(segment)

        return segments

    def create_space(self, space_type: str, name: str) -> Dict:
        """
        Create a Space for organizing concepts.

        Args:
            space_type: Type of space (conceptual, visual, temporal)
            name: Space name

        Returns:
            Space data structure
        """
        space_type_map = {
            'conceptual': 1,
            'visual': 2,
            'temporal': 3
        }

        space = {
            'type': space_type_map.get(space_type.lower(), 1),
            'name': name
        }

        return space

    def generate_world(self, task: str, concept_graph: Dict) -> Dict:
        """
        Generate a complete conceptual world from task and concepts.

        Args:
            task: Task description
            concept_graph: Extracted concept graph

        Returns:
            Complete world data structure
        """
        # Create the main conceptual space
        space = self.create_space('conceptual', f"World: {task[:50]}")

        # Create the central singularity
        singularity = self.create_singularity(task, concept_graph)

        # Create points from concepts
        points = self.create_points(concept_graph)

        # Create segments (connections)
        segments = self.create_segments(concept_graph, points)

        # Create objects from entities
        objects = self._create_objects_from_graph(concept_graph)

        # Create relations between objects
        relations = self._create_relations(concept_graph, objects)

        world = {
            'space': space,
            'singularity': singularity,
            'points': points,
            'segments': segments,
            'objects': objects,
            'relations': relations,
            'metadata': {
                'total_concepts': len(points),
                'total_connections': len(segments),
                'total_objects': len(objects),
                'dimensions': self.dimensions,
                'layers': singularity['layers_count']
            }
        }

        return world

    def _calculate_layers(self, concept_graph: Dict) -> int:
        """Calculate number of layers based on graph complexity."""
        num_nodes = len(concept_graph.get('nodes', []))

        if num_nodes < 5:
            return 1
        elif num_nodes < 15:
            return 2
        elif num_nodes < 30:
            return 3
        else:
            return 4

    def _generate_visual_representation(self, concept_graph: Dict) -> Dict:
        """Generate visual representation data for the concept graph."""
        return {
            'graph_structure': 'multidimensional',
            'node_count': len(concept_graph.get('nodes', [])),
            'edge_count': len(concept_graph.get('edges', [])),
            'layout': 'force_directed'
        }

    def _text_to_vector(self, text: str) -> List[float]:
        """
        Convert text to a simple heuristic vector.
        Uses character frequencies and text statistics.
        """
        if not text:
            return [0.0] * 10

        # Calculate various text features
        features = [
            len(text) / 100.0,  # Length normalized
            len(text.split()) / 20.0,  # Word count normalized
            sum(1 for c in text if c.isupper()) / max(len(text), 1),  # Capital letter ratio
            sum(1 for c in text if c.isdigit()) / max(len(text), 1),  # Digit ratio
            sum(1 for c in text if c.isspace()) / max(len(text), 1),  # Space ratio
            len(set(text.lower())) / max(len(text), 1),  # Unique char ratio
            text.count('а') / max(len(text), 1),  # Frequency of 'а'
            text.count('e') / max(len(text), 1),  # Frequency of 'e'
            hash(text) % 100 / 100.0,  # Hash-based feature
            len(text) % 10 / 10.0  # Length modulo
        ]

        return features[:10]

    def _generate_position_vector(self, node: Dict) -> List[float]:
        """
        Generate position vector for a node in multidimensional space.
        Uses heuristic placement based on node properties.
        """
        # Use hash of node ID for deterministic but distributed placement
        node_id = node.get('id', '')
        hash_value = hash(node_id)

        position = []
        for i in range(self.dimensions):
            # Generate coordinate using different hash seeds
            seed = hash_value + i * 1000
            coord = (seed % 10000) / 10000.0  # Normalize to [0, 1]

            # Apply weight influence
            weight = node.get('weight', 0.5)
            coord = coord * weight

            position.append(coord)

        return position

    def _determine_point_type(self, node: Dict) -> int:
        """
        Determine point type: 1 - resource, 2 - content, 3 - semantic
        """
        node_type = node.get('type', 'keyword')

        type_mapping = {
            'keyword': 3,  # Semantic
            'concepts': 3,  # Semantic
            'actions': 2,  # Content
            'attributes': 1  # Resource
        }

        return type_mapping.get(node_type, 3)

    def _calculate_point_layer(self, node: Dict, concept_graph: Dict) -> int:
        """Calculate layer for a point based on its connectivity."""
        node_id = node.get('id')
        edges = concept_graph.get('edges', [])

        # Count connections
        connections = sum(1 for e in edges if e.get('from') == node_id or e.get('to') == node_id)

        # Determine layer based on connectivity
        if connections == 0:
            return 0
        elif connections <= 2:
            return 1
        elif connections <= 5:
            return 2
        else:
            return 3

    def _generate_point_visualization(self, node: Dict, position: List[float]) -> Dict:
        """Generate visualization data for a point."""
        return {
            'position': position,
            'size': node.get('weight', 1.0) * 10,
            'color': self._calculate_color_from_position(position),
            'label': node.get('id', '')
        }

    def _calculate_color_from_position(self, position: List[float]) -> str:
        """Calculate color based on position in space."""
        if len(position) >= 3:
            r = int(position[0] * 255)
            g = int(position[1] * 255)
            b = int(position[2] * 255)
        else:
            r = int(position[0] * 255) if len(position) > 0 else 128
            g = int(position[0] * 128) if len(position) > 0 else 128
            b = 200
        return f"#{r:02x}{g:02x}{b:02x}"

    def _encode_segment_type(self, type_name: str) -> int:
        """Encode segment type as integer."""
        type_map = {
            'is_a': 1,
            'has': 2,
            'creates': 3,
            'related_to': 4,
            'via': 5,
            'co_occurrence': 6
        }
        return type_map.get(type_name, 0)

    def _create_objects_from_graph(self, concept_graph: Dict) -> List[Dict]:
        """Create objects from concept graph nodes."""
        objects = []

        for node in concept_graph.get('nodes', []):
            if node.get('type') in ['concepts', 'keyword']:
                obj = {
                    'is_subject': node.get('weight', 0) > 0.7,
                    'name': node.get('id', ''),
                    'type': node.get('type', 'concept'),
                    'data': node
                }
                objects.append(obj)

        return objects

    def _create_relations(self, concept_graph: Dict, objects: List[Dict]) -> List[Dict]:
        """Create relations between objects."""
        relations = []
        object_name_map = {obj['name']: i for i, obj in enumerate(objects)}

        for edge in concept_graph.get('edges', []):
            from_name = edge.get('from')
            to_name = edge.get('to')

            if from_name in object_name_map and to_name in object_name_map:
                relation = {
                    'type': edge.get('type', 'related'),
                    'from_object_idx': object_name_map[from_name],
                    'to_object_idx': object_name_map[to_name],
                    'weight': edge.get('weight', 1.0)
                }
                relations.append(relation)

        return relations

File: dotnet/src/MetatronAGI.Core/MetaModelBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;

namespace MetatronAGI.Core
{
    /// <summary>
    /// Meta-Model Builder - Heuristic-based world generation
    /// Creates meta-models and conceptual worlds from extracted concepts
    /// Uses graph algorithms and procedural generation without LLMs
    /// </summary>
    public class MetaModelBuilder
    {
        private readonly int _dimensions;
        private readonly Dictionary<string, List<double>> _conceptPositions;
        private readonly List<int> _layers;

        public MetaModelBuilder(int dimensions = 3)
        {
            _dimensions = dimensions;
            _conceptPositions = new Dictionary<string, List<double>>();
            _layers = new List<int>();
        }

        public Singularity CreateSingularity(string task, ConceptGraph conceptGraph)
        {
            // Calculate layers based on concept complexity
            var layersCount = CalculateLayers(conceptGraph);

            // Generate visual representation
            var visualRep = GenerateVisualRepresentation(conceptGraph);

            return new Singularity
            {
                Task = task,
                TaskVector = TextToVector(task),
                LayersCount = layersCount,
                VisualRepresentation = visualRep,
                Content = JsonSerializer.Serialize(conceptGraph),
                Version = 1
            };
        }

        public List<Point> CreatePoints(ConceptGraph conceptGraph, int? singularityId = null)
        {
            var points = new List<Point>();

            foreach (var node in conceptGraph.Nodes)
            {
                // Generate position vector in multidimensional space
                var positionVector = GeneratePositionVector(node);

                // Determine point type based on node characteristics
                var pointType = DeterminePointType(node);

                // Calculate layer based on node weight and connections
                var layer = CalculatePointLayer(node, conceptGraph);

                var point = new Point
                {
                    Task = node.Id,
                    TaskVector = TextToVector(node.Id),
                    PositionVector = positionVector,
                    Layer = layer,
                    SpaceDimension = _dimensions,
                    Name = node.Id,
                    Content = JsonSerializer.Serialize(node),
                    Weight = node.Weight,
                    Type = pointType,
                    IsSuperCluster = node.Weight > 0.8,
                    VisualRepresentation = GeneratePointVisualization(node, positionVector),
                    SingularityId = singularityId
                };

                points.Add(point);
                _conceptPositions[node.Id] = positionVector;
            }

            return points;
        }

        public List<Segment> CreateSegments(ConceptGraph conceptGraph, List<Point> points)
        {
            var segments = new List<Segment>();
            var pointIdMap = points.Select((p, i) => new { p.Name, Index = i })
                                  .ToDictionary(x => x.Name, x => x.Index);

            foreach (var edge in conceptGraph.Edges)
            {
                if (pointIdMap.ContainsKey(edge.From) && pointIdMap.ContainsKey(edge.To))
                {
                    segments.Add(new Segment
                    {
                        TypeName = edge.Type,
                        Type = EncodeSegmentType(edge.Type),
                        Weight = edge.Weight,
                        FromPointIdx = pointIdMap[edge.From],
                        ToPointIdx = pointIdMap[edge.To]
                    });
                }
            }

            return segments;
        }

        public Space CreateSpace(string spaceType, string name)
        {
            var spaceTypeMap = new Dictionary<string, int>
            {
                ["conceptual"] = 1,
                ["visual"] = 2,
                ["temporal"] = 3
            };

            return new Space
            {
                Type = spaceTypeMap.ContainsKey(spaceType.ToLower())
                    ? spaceTypeMap[spaceType.ToLower()]
                    : 1,
                Name = name
            };
        }

        public World GenerateWorld(string task, ConceptGraph conceptGraph)
        {
            // Create the main conceptual space
            var space = CreateSpace("conceptual", $"World: {task.Substring(0, Math.Min(50, task.Length))}");

            // Create the central singularity
            var singularity = CreateSingularity(task, conceptGraph);

            // Create points from concepts
            var points = CreatePoints(conceptGraph);

            // Create segments (connections)
            var segments = CreateSegments(conceptGraph, points);

            // Create objects from entities
            var objects = CreateObjectsFromGraph(conceptGraph);

            // Create relations between objects
            var relations = CreateRelations(conceptGraph, objects);

            return new World
            {
                Space = space,
                Singularity = singularity,
                Points = points,
                Segments = segments,
                Objects = objects,
                Relations = relations,
                Metadata = new WorldMetadata
                {
                    TotalConcepts = points.Count,
                    TotalConnections = segments.Count,
                    TotalObjects = objects.Count,
                    Dimensions = _dimensions,
                    Layers = singularity.LayersCount
                }
            };
        }

        private int CalculateLayers(ConceptGraph conceptGraph)
        {
            var numNodes = conceptGraph.Nodes.Count;

            if (numNodes < 5) return 1;
            if (numNodes < 15) return 2;
            if (numNodes < 30) return 3;
            return 4;
        }

        private Dictionary<string, object> GenerateVisualRepresentation(ConceptGraph conceptGraph)
        {
            return new Dictionary<string, object>
            {
                ["graph_structure"] = "multidimensional",
                ["node_count"] = conceptGraph.Nodes.Count,
                ["edge_count"] = conceptGraph.Edges.Count,
                ["layout"] = "force_directed"
            };
        }

        private List<double> TextToVector(string text)
        {
            if (string.IsNullOrEmpty(text))
                return Enumerable.Repeat(0.0, 10).ToList();

            var features = new List<double>
            {
                text.Length / 100.0,
                text.Split().Length / 20.0,
                text.Count(c => char.IsUpper(c)) / (double)Math.Max(text.Length, 1),
                text.Count(c => char.IsDigit(c)) / (double)Math.Max(text.Length, 1),
                text.Count(c => char.IsWhiteSpace(c)) / (double)Math.Max(text.Length, 1),
                text.ToLower().Distinct().Count() / (double)Math.Max(text.Length, 1),
                text.Count(c => c == 'а') / (double)Math.Max(text.Length, 1),
                text.Count(c => c == 'e') / (double)Math.Max(text.Length, 1),
                (text.GetHashCode() % 100) / 100.0,
                (text.Length % 10) / 10.0
            };

            return features.Take(10).ToList();
        }

        private List<double> GeneratePositionVector(ConceptNode node)
        {
            var nodeId = node.Id;
            var hashValue = nodeId.GetHashCode();
            var position = new List<double>();

            for (int i = 0; i < _dimensions; i++)
            {
                var seed = hashValue + i * 1000;
                var coord = (Math.Abs(seed % 10000)) / 10000.0;

                // Apply weight influence
                coord *= node.Weight;

                position.Add(coord);
            }

            return position;
        }

        private int DeterminePointType(ConceptNode node)
        {
            var typeMapping = new Dictionary<string, int>
            {
                ["keyword"] = 3,
                ["concepts"] = 3,
                ["actions"] = 2,
                ["attributes"] = 1
            };

            return typeMapping.ContainsKey(node.Type) ? typeMapping[node.Type] : 3;
        }

        private int CalculatePointLayer(ConceptNode node, ConceptGraph conceptGraph)
        {
            var nodeId = node.Id;
            var connections = conceptGraph.Edges.Count(e => e.From == nodeId || e.To == nodeId);

            if (connections == 0) return 0;
            if (connections <= 2) return 1;
            if (connections <= 5) return 2;
            return 3;
        }

        private Dictionary<string, object> GeneratePointVisualization(ConceptNode node, List<double> position)
        {
            return new Dictionary<string, object>
            {
                ["position"] = position,
                ["size"] = node.Weight * 10,
                ["color"] = CalculateColorFromPosition(position),
                ["label"] = node.Id
            };
        }

        private string CalculateColorFromPosition(List<double> position)
        {
            int r, g, b;

            if (position.Count >= 3)
            {
                r = (int)(position[0] * 255);
                g = (int)(position[1] * 255);
                b = (int)(position[2] * 255);
            }
            else
            {
                r = position.Count > 0 ? (int)(position[0] * 255) : 128;
                g = position.Count > 0 ? (int)(position[0] * 128) : 128;
                b = 200;
            }

            return $"#{r:X2}{g:X2}{b:X2}";
        }

        private int EncodeSegmentType(string typeName)
        {
            var typeMap = new Dictionary<string, int>
            {
                ["is_a"] = 1,
                ["has"] = 2,
                ["creates"] = 3,
                ["related_to"] = 4,
                ["via"] = 5,
                ["co_occurrence"] = 6
            };

            return typeMap.ContainsKey(typeName) ? typeMap[typeName] : 0;
        }

        private List<WorldObject> CreateObjectsFromGraph(ConceptGraph conceptGraph)
        {
            var objects = new List<WorldObject>();

            foreach (var node in conceptGraph.Nodes)
            {
                if (node.Type == "concepts" || node.Type == "keyword")
                {
                    objects.Add(new WorldObject
                    {
                        IsSubject = node.Weight > 0.7,
                        Name = node.Id,
                        Type = node.Type,
                        Data = node
                    });
                }
            }

            return objects;
        }

        private List<Relation> CreateRelations(ConceptGraph conceptGraph, List<WorldObject> objects)
        {
            var relations = new List<Relation>();
            var objectNameMap = objects.Select((obj, i) => new { obj.Name, Index = i })
                                      .ToDictionary(x => x.Name, x => x.Index);

            foreach (var edge in conceptGraph.Edges)
            {
                if (objectNameMap.ContainsKey(edge.From) && objectNameMap.ContainsKey(edge.To))
                {
                    relations.Add(new Relation
                    {
                        Type = edge.Type,
                        FromObjectIdx = objectNameMap[edge.From],
                        ToObjectIdx = objectNameMap[edge.To],
                        Weight = edge.Weight
                    });
                }
            }

            return relations;
        }
    }

    // Data models
    public class Singularity
    {
        public string Task { get; set; } = string.Empty;
        public List<double> TaskVector { get; set; } = new List<double>();
        public int LayersCount { get; set; }
        public Dictionary<string, object> VisualRepresentation { get; set; } = new Dictionary<string, object>();
        public string Content { get; set; } = string.Empty;
        public int Version { get; set; }
    }

    public class Point
    {
        public string Task { get; set; } = string.Empty;
        public List<double> TaskVector { get; set; } = new List<double>();
        public List<double> PositionVector { get; set; } = new List<double>();
        public int Layer { get; set; }
        public int SpaceDimension { get; set; }
        public string Name { get; set; } = string.Empty;
        public string Content { get; set; } = string.Empty;
        public double Weight { get; set; }
        public int Type { get; set; }
        public bool IsSuperCluster { get; set; }
        public Dictionary<string, object> VisualRepresentation { get; set; } = new Dictionary<string, object>();
        public int? SingularityId { get; set; }
    }

    public class Segment
    {
        public string TypeName { get; set; } = string.Empty;
        public int Type { get; set; }
        public double Weight { get; set; }
        public int FromPointIdx { get; set; }
        public int ToPointIdx { get; set; }
    }

    public class Space
    {
        public int Type { get; set; }
        public string Name { get; set; } = string.Empty;
    }

    public class WorldObject
    {
        public bool IsSubject { get; set; }
        public string Name { get; set; } = string.Empty;
        public string Type { get; set; } = string.Empty;
        public object Data { get; set; } = new object();
    }

    public class Relation
    {
        public string Type { get; set; } = string.Empty;
        public int FromObjectIdx { get; set; }
        public int ToObjectIdx { get; set; }
        public double Weight { get; set; }
    }

    public class World
    {
        public Space Space { get; set; } = new Space();
        public Singularity Singularity { get; set; } = new Singularity();
        public List<Point> Points { get; set; } = new List<Point>();
        public List<Segment> Segments { get; set; } = new List<Segment>();
        public List<WorldObject> Objects { get; set; } = new List<WorldObject>();
        public List<Relation> Relations { get; set; } = new List<Relation>();
        public WorldMetadata Metadata { get; set; } = new WorldMetadata();
    }

    public class WorldMetadata
    {
        public int TotalConcepts { get; set; }
        public int TotalConnections { get; set; }
        public int TotalObjects { get; set; }
        public int Dimensions { get; set; }
        public int Layers { get; set; }
    }
}
